### PR TITLE
Concept for v2 dplyr verbs 

### PR DESCRIPTION
**Pull request by elinw**
This is how I would propose to handle the new dplyr.  Let's force users who have the old version of dplyr to get the same classes as users of the new dplyr get (which is to say force them to have skim_df as a class).  This makes more sense and is forward looking (and eventually we can drop support for old dplyr).  That way we have the breaking change along with all the other breaking changes.

### CODE REVIEW

**Comment by elinw** [comment]
I'm going to go though a number of issues addressed by this PR and outline still remaining issues.  
There is a basic issue which is that dplyr changed APIs and that has an impact on skimr. Prior to the upcoming dplyr 0.8.0 using dplyr verbs would strip classes from the `.data`.  We agree with the dplyr team that this was not the correct behavior and that it was a surprise to users.  However, this change of API means that users who have different versions of dplyr installed will get different results for printing the skim_df data.  The new version prints using the skimr print function which is preferable.    

What we decided to do is to support older versions of dplyr by forcing the class and other attributes to be retained when using the v2 version of skimr.  (For v1 we took the opposite approach of retaining the old behavior.)

This also forced us to consider more deeply the impact of the new v2 version of skim_df, which is a wide data frame/tibble object.  For example, we realized that if we retain the skim_df class we must make sure that the metadata columns `type` and `variable` are present, otherwise the skim print cannot work. 

Key to this is @michaelquinn32 's commit of the `rebuild_skim_object()` internal function.   However for verbs like `select` and `transmute` we also had to add handling to preserve the two metadata columns. 

At some point in the future we an drop support for the older versions of dplyr and simplify some of this, but since 0.8.0 has breaking changes that will potentially impact some users workflows we'll hold off on that.  We will need to retain some special handling indefinitely because of the metadata issue.

This PR generalizes the code Michael wrote for `mutate`.  However, it always returns a skim_df object.  Users who really want to have another type of object can use `as.data.frame` as one option.  

At this stage this is still looking for feedback and comments about whether this approach is the best one.  Also documentation needs more work and there is a one bug that became visible here which can be seen by running the code in the passing tests for  transmute and mutate which is that the new variable is only calculated for one type.  However, I believe that issue is unrelated.

File: R/dplyr.R [hunk 1]
```diff
@@ -25,3 +26,47 @@ mutate.skim_df <- function(.data, ...) {
...
+}
+
+#' Use dplyr verb select on skim_df objects.
+#' Always includes variable and type columns even if user has not selected them.
+#' @seealso [`dplyr::select()`]
+#' @inheritParams dplyr::select
+#' @return  skim_df object
+#' @export
+select.skim_df <-function (.data, ...) {
+   mc <- match.call()
```
Comment (michaelquinn32) [inline R/dplyr.R]: This behavior should be covered by our package's function "focus".

To be safe, we could still keep this around, only matching the attributes. Like transmute above, you could warn the user that they were destroying the skim object and revert it to df.

File: R/dplyr.R [hunk 2]
```diff
@@ -25,3 +26,47 @@ mutate.skim_df <- function(.data, ...) {
...
+  transmuted <- transmuted[!colnames(transmuted) %in% c("variable", "type")]
+  base_skim <- .data[c("variable", "type")]
+  transmuted <- cbind(base_skim, transmuted)
+  rebuild_skim_obj(transmuted, .data)
+}
+
+#' Use dplyr verb select on skim_df objects.
+#' Always includes variable and type columns even if user has not selected them.
+#' @seealso [`dplyr::select()`]
+#' @inheritParams dplyr::select
```
Comment (michaelquinn32) [inline R/dplyr.R]: inheriting the params ended up referencing a bunch of packages like dbplyr that I didn't want. This would be better if its explicit. This might also help with the DESCRIPTION insanity we're getting during travis checks.

File: R/dplyr.R [hunk 3]
```diff
@@ -25,3 +26,47 @@ mutate.skim_df <- function(.data, ...) {
...
+#' @export
+transmute.skim_df <- function(.data, ...) {
+  transmuted <- NextMethod("transmute")
+  transmuted <- transmuted[!colnames(transmuted) %in% c("variable", "type")]
+  base_skim <- .data[c("variable", "type")]
+  transmuted <- cbind(base_skim, transmuted)
+  rebuild_skim_obj(transmuted, .data)
+}
+
+#' Use dplyr verb select on skim_df objects.
```
Comment (michaelquinn32) [inline R/dplyr.R]:
```suggest
#' Use dplyr verb select on skim_df objects.
#'
```

Comment (michaelquinn32) [reply R/dplyr.R]: There should be an empty line after the title and another after the description.

File: R/dplyr.R [hunk 4]
```diff
@@ -25,3 +26,47 @@ mutate.skim_df <- function(.data, ...) {
...
+#'   name of a new variable, and the value will be its corresponding value. 
+#'   Use `NULL` value in `mutate` to drop a variable.  New variables overwrite
+#'   existing variables of the same name.
+#'   @return A `skim_df` object, which also inherits the class(es) of the input
+#'   data. The `variable`` and `type` variables will be included
+#'   even if not named. 
+#'   In many ways, the object behaves like a [tibble::tibble()].
+#' @seealso [dplyr::transmute()] for the function's expected behavior.
+#' @export
+transmute.skim_df <- function(.data, ...) {
```
Comment (michaelquinn32) [inline R/dplyr.R]: I don't like changing the way transmute behaves within our package. My impression is that users will be much more familiar with transmute in dplyr than in skim.

Instead, we could warn the user that they are dropping relevant skim attributes and return a tibble always.

It would be nice to have a check whether an object `is_skim_df`:

```
is_skim_df <- function(object) {
  # has our class
  # has our columns
  # has our attributes
}

```

File: R/dplyr.R [hunk 5]
```diff
@@ -25,3 +26,47 @@ mutate.skim_df <- function(.data, ...) {
...
+#' features of a skim_df object.
+#'
+#' @param .data A `skim_df`, which behaves like a `tbl.`
+#' @param ... Name-value pairs of expressions, each with length 1 or the same
+#'   length as the number of rows in the group (if using [group_by()] or in the
+#'   entire input (if not using groups). The name of each argument will be the
+#'   name of a new variable, and the value will be its corresponding value. 
+#'   Use `NULL` value in `mutate` to drop a variable.  New variables overwrite
+#'   existing variables of the same name.
+#'   @return A `skim_df` object, which also inherits the class(es) of the input
```
Comment (michaelquinn32) [inline R/dplyr.R]: The spacing for the return tag appears off.

File: tests/testthat/test-dplyr.R [hunk 6]
```diff
@@ -1,35 +1,73 @@
...
 
 skimmed_iris <- skim(iris)
+nvar_iris <- length(iris)
+skimmers_used_iris <- attr(skimmed_iris, "skimmers_used")
+ntypes <- length(skimmers_used_iris)
+skimmers_iris <- union(skimmers_used_iris[[1]], skimmers_used_iris[[2]])
+length_skimmed_iris <- length(skimmers_iris) + 2
+classes_skimmed_iris <- class(skimmed_iris)
 
-test_that("dplyr::filter works as expected", {
```
Comment (michaelquinn32) [inline tests/testthat/test-dplyr.R]: I don't think you should revert the changes in the test.

Our issues with dplyr primarily emerge when we call `partition`. This happens when we print. The shape of the df isn't nearly as important and getting a printed result.

File: .travis.yml [hunk 7]
```diff
@@ -5,7 +5,7 @@ sudo: required
 warnings_are_errors: true
 env:
   global:
-    - NOT_CRAN=false
+    - NOT_CRAN=truee
```
Comment (michaelquinn32) [inline .travis.yml]: I think `truee` is a typo

File: R/dplyr.R (hunk 2 above)
Comment (elinw) [reply R/dplyr.R]: So you want to do the params explicitly if I understand you.

File: tests/testthat/test-dplyr.R (hunk 6 above)
Comment (elinw) [reply tests/testthat/test-dplyr.R]: Actually the shape a the key thing.  Filter takes a lot of forms, in one case you can filter based on `type`  and another you can filter based on a value of one of the calculated statistics and those have very different impacts on the printing.

File: R/dplyr.R (hunk 4 above)
Comment (michaelquinn32) [reply R/dplyr.R]: I'll handle that in a separate branch

File: R/dplyr.R (hunk 1 above)
Comment (michaelquinn32) [reply R/dplyr.R]: Well, we're still stuck then.

Basically select is now just an alias for focus, since they are doing essentially the same thing. If we're going this route, we shouldn't have two functions that behave the same way..

File: tests/testthat/test-dplyr.R (hunk 6 above)
Comment (michaelquinn32) [reply tests/testthat/test-dplyr.R]: I haven't seen an example of where filter modifies the columns and the rows. Do you have one?

File: R/dplyr.R (hunk 1 above)
Comment (elinw) [reply R/dplyr.R]: Correct,  I see three options.  Clearly we will not have two doing the same thing.  One can  allow selects that don't include variable and type,  but always (?) strip off the skim_df class.  One can prevent removal of variable and type and always keep the skim_df class.   I think a third option is have one function with a parameter that either allows you to or  prevents you from removing them.  Probably there are more possibilities but to me those are the main ones.  And then the question is which one do we call select and which do we call focus.

Comment (elinw) [reply R/dplyr.R]: If we don't have something called select and a users uses select, then they are by default allowed to not select variable and type (and any grouping variables!) but they also keep the skim_df class even if  it isn't a skim_df.   So I think at minimum we need a select method to deliver the message and strip off the class.  I don't think we want to sometimes return a skim_df class and other times not.

File: tests/testthat/test-dplyr.R (hunk 6 above)
Comment (elinw) [reply tests/testthat/test-dplyr.R]: Filter doesn't modify columns, just rows.  But it changes the print a lot if you filter based on type.

**Comment by elinw** [comment]
I'm going to close this since I think we've resolved the issues elsewhere.
